File: algorithms/src/Set/PowerSet.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Algorithms.Set
{
    public static class PowerSet
    {
        public static List<List<T>> GetPowerSet<T>(this IReadOnlyCollection<T> set)
        {
            List<List<T>> output = new List<List<T>>();

            if (set.Count == 0)
            {
                output.Add(new List<T>());
                return output;
            }

            T[] setCopy = set.ToArray();
            T setHead = setCopy[0];
            T[] setRest = setCopy.Skip(1).ToArray();

            foreach (List<T> powerSet in setRest.GetPowerSet())
            {
                List<T> buffer = new List<T>();
                buffer.Add(setHead);
                buffer.AddRange(powerSet);

                output.Add(buffer);
                output.Add(powerSet);
            }

            return output;
        }
    }
}

File: python/algorithms/others/power_set.py
from typing import List, TypeVar, Sequence

T = TypeVar('T')


def get_power_set(set_: Sequence[T]) -> List[List[T]]:
  output: List[List[T]] = []
  if len(set_) == 0:
    output.append([])
    return output
  set_head = set_[0]
  set_rest = set_[1:]
  for power_set in get_power_set(set_rest):
    buffer = [set_head]
    buffer.extend(power_set)
    output.append(buffer)
    output.append(power_set)
  return output
